src/sentry/incidents/logic.py

```diff
@@ -995,6 +1002,32 @@ def get_triggers_for_alert_rule(alert_rule):
     return AlertRuleTrigger.objects.filter(alert_rule=alert_rule)
 
 
+def trigger_incident_triggers(incident):
+    from sentry.incidents.tasks import handle_trigger_action
+
+    triggers = IncidentTrigger.objects.filter(incident=incident).select_related(
+        "alert_rule_trigger"
+    )
+    for trigger in triggers:
+        if trigger.status == TriggerStatus.ACTIVE.value:
+            with transaction.atomic():
+                method = "resolve"
+                for action in AlertRuleTriggerAction.objects.filter(
+                    alert_rule_trigger=trigger.alert_rule_trigger
+                ):
+                    for project in incident.projects.all():
+                        handle_trigger_action.apply_async(
+                            kwargs={
+                                "action_id": action.id,
+                                "incident_id": incident.id,
+                                "project_id": project.id,
+                                "method": method,
+                            },
+                        )
+                trigger.status = TriggerStatus.RESOLVED.value
+                trigger.save()
```

Comment:
It makes sense to wrap each trigger in a transaction, so that we either resolve the entire trigger and related actions or we don't. However, this adds some weirdness because we have a Celery task here, which is fired outside of the transaction.

Luckily there's a pattern that can help with this - we don't use it much but I want to start doing so since it'll make things less susceptible to race conditions. Basically, when firing your task, you can do:
`transaction.on_commit(handle_trigger_action.s(action_id=action.id, incident_id=incident.id, project_id=project.id, method=method).delay)`.
What this does is schedule the tasks to only be sent after the transaction is committed. This helps us avoid situations where maybe an exception is raised due to a network error and we roll back, but still fire out the actions.

Just some explanation on how this works: `handle_trigger_action.s` is an alias for `handle_trigger_action.signature`. This is basically like a partial for a celery task. So then when we pass `handle_trigger_action.s(<args>).delay` to `transaction.on_commit` this is a callback, which is called once the transaction is saved to the database. This does still leave a small window where this could fail (transaction commits, server process dies for some reason before we fire celery tasks), but it's not extremely likely.

Reply:
Will do